pytest_jupyterhub/jupyterhub_spawners.py

```diff
@@ -6,48 +6,60 @@
 import sys
 
 import pytest
-from jupyterhub.spawner import Spawner
 from jupyterhub.tests.mocking import MockHub
 from traitlets.config import Config
 
 
 @pytest.fixture
-async def spawner_class():
-    """Allows tests to setup the Spawner class that will be used by the hub."""
-    return Spawner
+async def get_configured_mockhub_instance():
+    """
+    Creates a MockHub instance from a provided traitlets.config Config object
+    or the empty Config is none is passed as an argument.
 
+    This fixture is a factory and returns a function.
+    More about factory fixtures at https://docs.pytest.org/en/latest/how-to/fixtures.html#factories-as-fixtures
 
-@pytest.fixture
-async def jupyterhub_spawner_class_config(spawner_class):
-    """Configures JupyterHub to use the Spawner class specified in the spawner_class fixture."""
-    spawner_class_config = {"JupyterHub": {"spawner_class": spawner_class}}
-    return Config(spawner_class_config)
+    It should be called like:
+    .. code-block:: python
+      def my_test(get_configured_mockhub_instance):
+         hub_instance = get_configured_mockhub_instance(config=Config({"a": "b"})
+         ...
+    """
 
+    def _create_configured_mockhub_instance(
+        config=Config(),
+    ):
+        MockHub.clear_instance()
+        mocked_app = MockHub.instance(config=config)
+        return mocked_app
 
-@pytest.fixture
-async def spawner_config():
-    """Allows tests to setup configurations that are specific to the spawner implementation"""
-    return Config()
+    return _create_configured_mockhub_instance
 
 
 @pytest.fixture
-async def app_for_spawners(request, jupyterhub_spawner_class_config, spawner_config):
-    """Creates an instance of a JupyterHub application that can be used by different spawner implementation test suites"""
-    final_config = jupyterhub_spawner_class_config
-    final_config.merge(spawner_config)
-    mocked_app = MockHub.instance(config=final_config)
-
-    def fin():
-        """Disconnects logging during cleanup because pytest closes captured FDs prematurely"""
-        mocked_app.log.handlers = []
-        MockHub.clear_instance()
-        try:
-            mocked_app.stop()
-        except Exception as e:
-            print("Error stopping Hub: %s" % e, file=sys.stderr)
+async def hub_app(hub_app_cleanup):
```

Comment:
Lastly, how does the `hub_app_cleanup` fixture end up doing the cleaning without being called or registered to perform a cleanup?

Reply:
The `hub_app_cleanup` fixture is [“requested”](https://docs.pytest.org/en/latest/how-to/fixtures.html#requesting-fixtures) by the `hub_app`. This means that every time a test requests the `hub_app`, the `hub_app_cleanup` is also ran.

But I believe it can be replaced with an [yield + teardown](https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended) inside the same fixture for clarity. Commit incoming.

Reply:
Yes!! The committed change has made it much clearer. Thank you.